reasoner/resolution/framework/Resolver.java

```diff
@@ -30,107 +29,59 @@
 
-    /*
-     *
-     * Handlers for messages sent into the execution actor that are dispatched via the actor model.
-     *
-     */
-    public void executeReceiveRequest(Request fromUpstream, ResolverRegistry registry) {
-        LOG.trace("{}: Receiving a new Request: {}", name, fromUpstream);
-        if (!isInitialised) {
-            LOG.debug(name + ": initialising downstream actors");
-            initialiseDownstreamActors(registry);
-            isInitialised = true;
-        }
-
-        ResponseProducer responseProducer = responseProducers.computeIfAbsent(fromUpstream, key -> {
-            LOG.debug("{}: Creating a new ResponseProducer for the given Request: {}", name, fromUpstream);
-            return createResponseProducer(fromUpstream);
-        });
-        Either<Request, Response> action = receiveRequest(fromUpstream, responseProducer);
-        if (action.isFirst()) requestFromDownstream(action.first(), fromUpstream, registry);
-        else respondToUpstream(action.second(), registry);
-    }
-
-    void executeReceiveAnswer(Response.Answer fromDownstream, ResolverRegistry registry) {
-        LOG.trace("{}: Receiving a new Answer from downstream: {}", name, fromDownstream);
-        Request sentDownstream = fromDownstream.sourceRequest();
-        Request fromUpstream = requestRouter.get(sentDownstream);
-        ResponseProducer responseProducer = responseProducers.get(fromUpstream);
-        Either<Request, Response> action = receiveAnswer(fromUpstream, fromDownstream, responseProducer);
-
-        if (action.isFirst()) requestFromDownstream(action.first(), fromUpstream, registry);
-        else respondToUpstream(action.second(), registry);
-    }
-
-    void executeReceiveExhausted(Response.Exhausted fromDownstream, ResolverRegistry registry) {
-        LOG.trace("{}: Receiving a new Exhausted from downstream: {}", name, fromDownstream);
-        Request sentDownstream = fromDownstream.sourceRequest();
-        Request fromUpstream = requestRouter.get(sentDownstream);
-        ResponseProducer responseProducer = responseProducers.get(fromUpstream);
-
-        Either<Request, Response> action = receiveExhausted(fromUpstream, fromDownstream, responseProducer);
-
-        if (action.isFirst()) requestFromDownstream(action.first(), fromUpstream, registry);
-        else respondToUpstream(action.second(), registry);
+    protected abstract ResponseProducer responseProducerReiterate(Request fromUpstream, ResponseProducer responseProducer, int newIteration);
 
+    protected Request fromUpstream(Request toDownstream) {
+        assert requestRouter.containsKey(toDownstream);
+        return requestRouter.get(toDownstream);
     }
 
-    /*
-     *
-     * Helper method private to this class.
-     *
-     * */
-    private void requestFromDownstream(Request request, Request fromUpstream, ResolverRegistry registry) {
+    protected void requestFromDownstream(Request request, Request fromUpstream, int iteration) {
```

Comment:
this and the following method are called by subclasses to do the "framework" bit of dispatching requests/responses and recording routing